### DISCUSSION TITLE
Tensor op to calculate viscous terms

### DISCUSSION THREAD

**Discussion by jagmohas** [AMReX-Codes/amrex#3205] [Q&A]
I IAMR (Diffusion.cpp, line 1647) Diffusion::getTensorViscTerms uses mlmg.apply to calculate the viscous terms. I can't figure out how it does so. Could you please suggest where to look for to understand this maths?

**Comment by WeiqunZhang** [comment]
Is this a programming question about what mlmg.apply does or a math question about what the viscous terms are?

**Reply by jagmohas** [reply]
Thanks Weiqun for looking into this. My question is how mmlg calculates the viscous terms. It could be mixed of both programming and math.  

I understand that with tensorop.setScalars(a, b) where a=0,b=-1,  we are solving (\nabla \cdot \beta \nabla ) \phi = f. I can't figure out where do we set f ? 

And when we call mlmg.apply({&visc_tmp}, {&s_tmp});, s_tmp is velocity? I understand that s_tmp contains the data from S assigned at line 1678 of Diffusion.cpp         

MultiFab::Copy(s_tmp,S,Xvel,0,AMREX_SPACEDIM,0);

**Reply by WeiqunZhang** [reply]
The `apply` function applies the linear operator `L` without involving the RHS of the linear system `L(phi) = f`.  More specifically the linear system in the code you mentioned is `alpha a v - beta div dot tau = rhs`, where `tau = eta [grad v + (grad v)^T] +  (kappa-(2/3)eta) (div v) I`, `eta` and `kappa` are shear and bulk viscosity, and I is identity tensor.  See `amrex/Src/LinearSolvers/MLMG/AMReX_MLTensorOp.H` for more details.

**Comment by drummerdoc** [comment]
The getViscTerms function is a little different though.  It is (1/Vol) Integral( Div(tau) ), i.e., does not have the time discretization terms that are usually the and is volume weighted, thus there's no a*alpha contribution, and b=-1 to get the sign right.
